### DISCUSSION TITLE
Using Next/Image and Markdown content

### DISCUSSION THREAD

**Discussion by tseidler** [vercel/next.js#18383] [Help] [answered]
I'm using Strapi as my CMS and am using a method very similar to the Next `cms-strapi` example to render my page content.

Using showdown instead of remark, but the general idea is the same. The markdown is converted to html and then added to the container:

https://github.com/vercel/next.js/blob/5d7f7a47afa554e5b04c15aebb7ee1f9d3628cfd/examples/blog-starter/lib/markdownToHtml.js#L5

And

https://github.com/vercel/next.js/blob/5d7f7a47afa554e5b04c15aebb7ee1f9d3628cfd/examples/blog-starter/components/post-body.js#L8

Now.. the next Next/Image sounds amazing, but is there a way to apply that to my 'content'? Most of my site larger images are actually in the content itself. Those are the hardest to optimise in my case, _not_ the more static images.

Any suggestions would be appreciated. I found https://github.com/vercel/next.js/discussions/16322, but that discussion is ~72 days old and has 0 replies ;-)

**Comment by leerob** [comment, marked as answer]
You're essentially looking for a way to provide React components while parsing your Markdown. There are many solutions, but here's two to consider:

1. [`@next/mdx`](https://mdxjs.com/getting-started/next)
2. [markdown-to-jsx](https://github.com/probablyup/markdown-to-jsx)

With this, you can specific _what_ component you want to handle your images. Then, you can provide `next/image`.

**Reply by tseidler** [reply]
Doh - I should've thought of that. Thanks for the suggestion ;-)

Edit for future readers: https://mdxjs.com/getting-started/#working-with-components

**Reply by weyert** [reply]
I think it's better to use `react-markdown` instead of `markdown-to-jsx` as the former is more relaxed about the Markdown syntax. The latter won't render your content correctly all the time.

**Reply by leerob** [reply]
Do you have a specific example?

**Reply by stephent** [reply]
I'm struggling to work out how to pass image size from Strapi via a markdown content field to the `next/image` component.

Neither `react-markdown` nor `markdown-to-jsx` seem to like the `=wxh` image size syntax (e.g. `![Example image](https://example.com/image.png =250x350)` (see [here](https://stackoverflow.com/a/21242579/220287)).

I was thinking that if I can include width and height in the markdown I could then pass that through to a custom image renderer component that could set the size properties for `next/image`. However, so far, no luck.

Any suggestions?

**Reply by SalahAdDin** [reply]
As @leerob below, you have to wrap the image with a component with at least a certain height, I did it and it worked fine.

**Reply by stephent** [reply]
@SalahAdDin - that’s understood. The question is how to pass the image size via Markdown from Strapi (or equivalent). You need the size either for the Image or the div.

**Reply by SalahAdDin** [reply]
It is not possible to pass it from **Strapi** as far as I know.

**Reply by mckennapaul27** [reply]
Anybody ever find a solution to this?

**Reply by mckennapaul27** [reply]
I am also using Strapi and cannot find a way to get access to the width and height. I think I'll just go with using the img tag specifying width and height directly in markdown

**Reply by tseidler** [reply]
Yeah, that's what I'm doing as well, unfortunately.

**Reply by mckennapaul27** [reply]
Yeah it's a shame but it's the only way to use next/image

**Reply by SalahAdDin** [reply]
This should be notified on the Strapi repo, isn't it?

**Reply by leerob** [reply]
You can also use `layout="fill"` and put a `max-width` around the surrounding element, e.g. in a blog post the max-width of the post might be 700px.

https://nextjs.org/docs/api-reference/next/image#layout

**Reply by mckennapaul27** [reply]
Thanks @leerob that's correct and certainly an option. The thing is, I am using a blur placeholder from cloudinary so I need to access the width and height at some point

```
    return (
        <div
            style={{
                maxWidth: `${width}px`,
                display: 'block',
                paddingBottom: '26px',
            }}>
            <div
                style={{
                    position: 'relative',
                    height: 0,
                    paddingTop: `${(height / width) * 100}%`,
                    backgroundImage: `url(${urlBlurred})`,
                    backgroundPosition: 'center center',
                    backgroundSize: '100%',
                }}>
                <div
                    style={{
                        position: 'absolute',
                        top: 0,
                        left: 0,
                        height: '100%',
                        width: '100%',
                    }}>
                    <div
                        style={{
                            // need this extra wrapper so that we provide Image with relative parent
                            position: 'relative',
                            height: '100%',
                            width: '100%',
                        }}>
                        <Image src={url} alt={alt} layout='fill' />
                    </div>
                </div>
            </div>
        </div>
    );
```

**Reply by florianwalther-private** [reply]
You can add the image width and height as URL parameters in markdown and then just read it in your markdown parser. Using react-markdown, my code looks like this:

```
<ReactMarkdown
    remarkPlugins={...}
    rehypePlugins={...}
    components={{
        img: ({ node, ...props }) => {
            const imageUrl = new URL(props.src!, process.env.NEXT_PUBLIC_TUTHUB_WEBSITE_URL);
            const queryParams = new URLSearchParams(imageUrl.search)
            const width = queryParams.get('width');
            const height = queryParams.get('height');

            return (
                <span className={styles.inPostImageWrapper}>
                    <Image
                        src={props.src!}
                        alt={props.alt}
                        width={width!}
                        height={height!}
                    />
                </span>
            )
        }
    }}
>
    {content}
</ReactMarkdown>
``` 

Markdown:

```
![Example image](/images/blog/my_example_image.png?width=500&height=400)
```

This transforms all images in my markdown into next/image and uses the dimensions from the URL 👍

**Reply by SalahAdDin** [reply]
> You can add the image width and height as URL parameters in markdown and then just read it in your markdown parser. Using react-markdown, my code looks like this:
> 
> ```
> <ReactMarkdown
>     remarkPlugins={...}
>     rehypePlugins={...}
>     components={{
>         img: ({ node, ...props }) => {
>             const imageUrl = new URL(props.src!, process.env.NEXT_PUBLIC_TUTHUB_WEBSITE_URL);
>             const queryParams = new URLSearchParams(imageUrl.search)
>             const width = queryParams.get('width');
>             const height = queryParams.get('height');
> 
>             return (
>                 <span className={styles.inPostImageWrapper}>
>                     <Image
>                         src={props.src!}
>                         alt={props.alt}
>                         width={width!}
>                         height={height!}
>                     />
>                 </span>
>             )
>         }
>     }}
> >
>     {content}
> </ReactMarkdown>
> ```
> 
> Markdown:
> 
> ```
> ![Example image](/images/blog/my_example_image.png?width=500&height=400)
> ```
> 
> This transforms all images in my markdown into next/image and uses the dimensions from the URL 👍

Let's say I'm getting the `URL` from **Cloudinary** through **Strapi CMS**, in that case, how can I get the **witdh** and **height**?

**Reply by florianwalther-private** [reply]
@SalahAdDin I don't know, sorry!

**Reply by florianwalther-private** [reply]
@SalahAdDin I have no experience with Strapi but could you maybe put the dimensions into the file name directly and parse them from there? Just an idea if you can't use URL params.

**Reply by dietrichmax** [reply]
@SalahAdDin You can get widths and heights of images with the url by querying the content api like
```js
`${STRAPI_API_URL}/upload/files?_url=${src}`
```

**Reply by vikasswebdev** [reply]
Thank you

**Comment by mikecann** [comment]
I too am struggling with this.

Im not too sure where this JSX should be placed, we still want things to be statically rendered. How do we get the image widths and heights for the Image component?

Would love a little example, to help understand. Theres probably quite a few people trying to use the new Next Image component with Markdown..

**Reply by gmichael89** [reply]
I wrote a [blog post](https://theviewport.io/post/using-nextjs-and-nextimage-with-mdx-markdown-processing) detailing how to provide `next/image` for markdown images by looking into the image's AST properties - but bare in mind that it's quite a contrived example. 

It might point you in the right direction as there are other properties on the image's AST node that you are able to access in the renderer

**Reply by SalahAdDin** [reply]
@gmichael89 I saw it and it specifies a width and height for all images, which is not what i want for my images, is there any other option to do that?

**Reply by leerob** [reply]
You have to put a width and height _somewhere_ if you want to replace your layout from shifting with next/image. If you defer from using it on the component, and use `layout=fill`, the component that wraps it will still need to set a width and a height.

**Reply by SalahAdDin** [reply]
Setting up a fixed thumbnail size is easy, let's say, for grids, for example; but when your content is dynamic and it comes from **markdown**, having a fixed size is not a convenient solution I think.

**Reply by florianwalther-private** [reply]
@sambecker Take a look at my comment in the other answer. You can put the image dimensions as URL params into the image URL.
